src/System.Net.NameResolution/src/System/Net/NameResolutionPal.Windows.cs

```diff
@@ -385,10 +387,206 @@ public static void EnsureSocketsAreInitialized()
+                    {
+                        if (socketAddress.Length == SocketAddressPal.IPv6AddressSize)
+                            addresses.Add(CreateIPv6Address(socketAddress));
+                    }
+
+                    result = result->ai_next;
+                }
+
+                if (canonicalName == null)
+                    canonicalName = state.HostName;
+
+                state.CompleteAsyncResult(new IPHostEntry
+                {
+                    HostName = canonicalName,
+                    Aliases = Array.Empty<string>(),
+                    AddressList = addresses.ToArray()
+                });
+            }
+            finally
+            {
+                GetAddrInfoExContext.FreeContext(context);
+            }
+        }
+
+        private static unsafe IPAddress CreateIPv4Address(ReadOnlySpan<byte> socketAddress)
+        {
+            long address = (long)SocketAddressPal.GetIPv4Address(socketAddress) & 0x0FFFFFFFF;
+            return new IPAddress(address);
+        }
+
+        private static unsafe IPAddress CreateIPv6Address(ReadOnlySpan<byte> socketAddress)
+        {
+            Span<byte> address = stackalloc byte[IPAddressParserStatics.IPv6AddressBytes];
+            uint scope;
+            SocketAddressPal.GetIPv6Address(socketAddress, address, out scope);
+
+            return new IPAddress(address, (long)scope);
+        }
+
+        #region GetAddrInfoAsync Helper Classes
+
+        //
+        // Warning: If this ever ported to NETFX, AppDomain unloads needs to be handled
+        // to protect against AppDomainUnloadException if there are pending operations.
+        //
+
+        private sealed class GetAddrInfoExState
+        {
+            private DnsResolveAsyncResult _asyncResult;
+            private object _result;
+
+            public string HostName => _asyncResult.HostName;
+
+            public GetAddrInfoExState(DnsResolveAsyncResult asyncResult)
+            {
+                _asyncResult = asyncResult;
+            }
+
+            public void CompleteAsyncResult(object o)
+            {
+                // We don't want to expose the GetAddrInfoEx callback thread to user code.
+                // The callback occurs in a native windows thread pool.
+
+                _result = o;
+
+                Task.Factory.StartNew(s =>
+                {
+                    var self = (GetAddrInfoExState)s;
+                    self._asyncResult.InvokeCallback(self._result);
+                }, this, CancellationToken.None, TaskCreationOptions.DenyChildAttach, TaskScheduler.Default);
+            }
+
+            public IntPtr CreateHandle()
+            {
+                GCHandle handle = GCHandle.Alloc(this, GCHandleType.Normal);
+                return GCHandle.ToIntPtr(handle);
+            }
+
+            public static GetAddrInfoExState FromHandle(IntPtr handle)
```

Comment:
Nit: FromHandleAndFree?  Might help make the cleanup more clear for someone reading the call site.